### DISCUSSION TITLE
Rendering canvas/video on the native side

### DISCUSSION THREAD

**Discussion by sztomi** [tauri-apps/wry#284] [Q&A]
I've been looking for a GUI solution for Rust to build a media player. The plan is to wrap libmpv for video playback. libmpv has a couple of ways to facilitate integration, but the best option is rendering via opengl: https://github.com/mpv-player/mpv-examples/blob/master/libmpv/README.md

Is wry capable of enabling that?

**Comment by evoactivity** [comment]
Would love some insight into how this could be acheived as I also want to build a media player style application.

**Reply by evoactivity** [reply]
I'm wondering if it would be possible to use @wusyong's [fork of glutin](https://github.com/wusyong/glutin) with a tauri window since tauri uses tao? I want to have a window containing an opengl background with a transparent webview rendering on top of that.

I'm a web dev rather than a rust dev, so if it is achievable I'd love a little guidance :)

**Reply by wusyong** [reply]
We have an experient branch on tauri but I didn't merge back yet.
I have another attempt which works better for gtk and I'll try to implement in next few weeks.

**Reply by prayag17** [reply]
Any updates on this?

**Reply by Ar-mane** [reply]
updates plz ?

**Reply by prayag17** [reply]
https://discord.com/channels/616186924390023171/1215223447693692958

I was able to get mpv to render throu command line in tauri window and also Gstreamer with Gstreamer rs, but to use Gstreamer I still need to use multiwebview feat from tauri v2 as Gstreamer was rendering above the surface and covering all html elements

**Reply by evoactivity** [reply]
@prayag17 Nice work! Could you please share some example code?

**Reply by prayag17** [reply]
I don't have my laptop on me for a couple of days but when I get it back, I will share the code. Tho it is pretty basic, I got the idea from https://github.com/fltk-rs/demos/blob/master/mpv/src/main.rs and just supplied the raw window handle from tauri's api in rust

**Reply by ALEZ-DEV** [reply]
@prayag17 Any update on this?

**Reply by jaruba** [reply]
@prayag17 i am also interested if you could share the code

**Comment by drakonkat** [comment]
Any news on this?

**Comment by dukeeagle** [comment]
Hey everyone! Wanted to bump this to see if there's been any progress

**Comment by felix307253927** [comment]
Any updates on this?

**Comment by evoactivity** [comment]
I asked this on the discord

> So now that we can create multiple webviews inside the window, and we have opengl through glutin. Is this question any closer to having a "yes, here's how to do it" answer?

and got this response from @FabianLars 

> i'd say yes. When the multi webview thing was being implemented there was also a wgpu rendering in that same window
ah lol, we even added a wgpu example https://github.com/tauri-apps/wry/blob/dev/examples/wgpu.rs
and since wry is now uncoupled from tao (aka you can use winit for example too) you probably don't even need the glutin fork

https://discord.com/channels/616186924390023171/731495028677148753/1174355856511488143

**Comment by MirazMac** [comment]
Any updates or example code regarding this?

**Comment by sakshatshinde** [comment]
Any update on this?

**Comment by scerelli** [comment]
I'm also interested in this

**Comment by jaruba** [comment]
mentioning my interest as well

**Comment by nini22P** [comment]
~~Hi everyone, I created a simple example program based on the above tips.~~

Update: I've developed the original example into two Tauri plugin, which uses JSON IPC or libmpv to communicate with the embedded mpv window. You can find it here:

https://github.com/nini22P/mpv-tauri

The OpenGL rendering implementation is not yet complete.

**Comment by nini22P** [comment]
Hello everyone, I've made some progress with OpenGL rendering, but it flickers when resizing the window.

https://github.com/tauri-apps/tauri/discussions/14228
